### PR TITLE
Use googleJobstore in provisioner tests.

### PR DESCRIPTION
**Pull request by ejacox**

### CODE REVIEW

**Comment by ejacox** [comment]
Jenkins, test this please

File: src/toil/utils/toilLaunchCluster.py [hunk 1]
```diff
@@ -84,14 +84,16 @@ def main():
         try:
             from toil.provisioners.aws.awsProvisioner import AWSProvisioner
         except ImportError:
-            raise RuntimeError('The aws extra must be installed to use this provisioner')
+            print('The aws extra must be installed to use this provisioner')
+            raise
```
Comment (jessebrennan) [inline src/toil/utils/toilLaunchCluster.py]: I don't disapprove, Just wondering why this change.

File: src/toil/test/provisioners/gceProvisionerTest.py [hunk 2]
```diff
@@ -287,7 +286,7 @@ def _runScript(self, toilOptions):
         # instances
         runCommand = ['/home/venv/bin/python', '/home/sort.py', '--fileToSort=/home/s3am/bin/asadmin', '--sortMemory=0.6G', '--mergeMemory=3.0G']
         runCommand.extend(toilOptions)
-        runCommand.append('--sseKey=/home/keyFile')
+        #runCommand.append('--sseKey=/home/keyFile')
```
Comment (jessebrennan) [inline src/toil/test/provisioners/gceProvisionerTest.py]: I guess google job store doesn't work with sse? Is there an issue for this?

File: src/toil/provisioners/gceProvisioner.py [hunk 3]
```diff
@@ -482,56 +487,81 @@ def addNodes(self, nodeType, numNodes, preemptable):
...
-                        else:
-                            logger.debug("Rsync to new nodes failed, trying again")
-                            retries += 1
-                            time.sleep(20*retries)
-            workersCreated += 1
-        logger.info('Launched %s new instance(s)', numNodes)
-        if len(instancesLaunched) != workersCreated:
-            logger.error("Failed to launch %d workers", len(instancesLaunched)-workersCreated)
+        # Try a few times to create the requested number of workers
+        while numNodes-workersCreated > 0 and retries < 3:
```
Comment (jessebrennan) [inline src/toil/provisioners/gceProvisioner.py]: This is a nice, robust retry mechanism. I like it 👍

File: src/toil/provisioners/gceProvisioner.py [hunk 4]
```diff
@@ -482,56 +487,81 @@ def addNodes(self, nodeType, numNodes, preemptable):
...
+                self.terminateNodes(failedWorkers)
+            retries += 1
+
+        logger.info('Launched %d new instance(s)', numNodes)
+        if numNodes != workersCreated:
+            logger.error("Failed to launch %d worker(s)", numNodes-workersCreated)
         return workersCreated
 
+
+    def _injectWorkerFiles(self, ip):
```
Comment (jessebrennan) [inline src/toil/provisioners/gceProvisioner.py]: These seem like much more robust ways of dealing with this. Definitely want to keep this model in mind for the provisioner refactor.

File: src/toil/provisioners/gceProvisioner.py [hunk 5]
```diff
@@ -852,12 +876,9 @@ def _waitForAppliance(cls, ip_address, role, keyName='core'):
                     logger.info('...Still waiting for appliance, trying again in %s sec...' % sleepTime)
                     time.sleep(sleepTime)
             except:
-                if retries == 5:
-                    raise # givup
```
Comment (jessebrennan) [inline src/toil/provisioners/gceProvisioner.py]: Why did you get rid of this?

File: src/toil/provisioners/gceProvisioner.py [hunk 6]
```diff
@@ -852,12 +876,9 @@ def _waitForAppliance(cls, ip_address, role, keyName='core'):
                     logger.info('...Still waiting for appliance, trying again in %s sec...' % sleepTime)
                     time.sleep(sleepTime)
             except:
```
Comment (jessebrennan) [inline src/toil/provisioners/gceProvisioner.py]: Now it will just eat up exception and keep retrying which is _not_ good behavior. What exception do we want to retry on?

File: src/toil/test/provisioners/gceProvisionerTest.py (hunk 2 above)
Comment (ejacox) [reply src/toil/test/provisioners/gceProvisionerTest.py]: SSE keys are not yet supported in Google. Partially because they are not tested. I pushed adding it to the refactor (see #2095).

File: src/toil/utils/toilLaunchCluster.py (hunk 1 above)
Comment (ejacox) [reply src/toil/utils/toilLaunchCluster.py]: It is clearer that way. Hiding exceptions by raising a different error is not a good practice.

File: src/toil/provisioners/gceProvisioner.py (hunk 4 above)
Comment (ejacox) [reply src/toil/provisioners/gceProvisioner.py]: I was thinking a decorator would look cleaner. Something for the refactor.

File: src/toil/provisioners/gceProvisioner.py (hunk 5 above)
Comment (ejacox) [reply src/toil/provisioners/gceProvisioner.py]: Time based failure seems to work better. That is covered by the loop on line 839. This mechanism seemed redundant or not very useful.

File: src/toil/provisioners/gceProvisioner.py (hunk 6 above)
Comment (ejacox) [reply src/toil/provisioners/gceProvisioner.py]: At this point, the connection works and docker is running. I don't know of any specific errors other than transient connection failures. What do you suggest?

Comment (jessebrennan) [reply src/toil/provisioners/gceProvisioner.py]: As far as I can tell, the only exception we would expect to encounter is from `_coreSSH` and is a `RuntimeError` so I think you should only retry on that exception or maybe only that exception and with a particular error code.

File: src/toil/utils/toilLaunchCluster.py [hunk 7]
```diff
@@ -84,14 +84,16 @@ def main():
...
         provisioner = AWSProvisioner()
     elif config.provisioner == 'gce':
         logger.info('Using a gce provisioner.')
         try:
             from toil.provisioners.gceProvisioner import GCEProvisioner
         except ImportError:
-            raise RuntimeError('The libCloud extra must be installed to use this provisioner')
+            print('The libCloud extra must be installed to use this provisioner')
```
Comment (jessebrennan) [inline src/toil/utils/toilLaunchCluster.py]: Somehow this slipped through the cracks... There is no `libCloud` extra! it should be google, right?

File: src/toil/provisioners/gceProvisioner.py [hunk 8]
```diff
@@ -351,13 +355,14 @@ def launchCluster(self, leaderNodeType, leaderSpotBid, nodeTypes, preemptableNod
...
+            or not self._copySshKeys(leader.public_ips[0], keyName)
+            or not self._injectFile(leader.public_ips[0], self.credentialsPath, GoogleJobStore.nodeServiceAccountJson,
+                                    'toil_leader')
+            or (self.botoPath and
+                not self._injectFile(leader.public_ips[0], self.botoPath, self.nodeBotoPath, 'toil_leader'))):
 
-        if self.botoPath:
-            self._rsyncNode(leader.public_ips[0], [self.botoPath, ':' + nodeBotoPath],
-                            applianceName='toil_leader')
+            raise RuntimeError("Failed to start leader")
```
Comment (DailyDreaming) [inline src/toil/provisioners/gceProvisioner.py]: I recall making this comment before, but I think each of these conditions should throw a different error message to relate to the user why it failed.

File: src/toil/utils/toilLaunchCluster.py (hunk 1 above)
Comment (DailyDreaming) [reply src/toil/utils/toilLaunchCluster.py]: I agree.  A plain `raise` is almost always better.

Comment (diekhans) [reply src/toil/utils/toilLaunchCluster.py]: Is this a program or a library?  print from libraries is generally not a good idea, they should go through the logger instead.

If this is a program, error messages should always be printed to stderr, not stdout.

I don't completely have the context here.  How are tests reporting error>

Comment (ejacox) [reply src/toil/utils/toilLaunchCluster.py]: Program. Stderr is better.

File: src/toil/utils/toilLaunchCluster.py (hunk 7 above)
Comment (DailyDreaming) [reply src/toil/utils/toilLaunchCluster.py]: Commented on these in the previous PR but they were not fixed.

Comment (ejacox) [reply src/toil/utils/toilLaunchCluster.py]: Try refreshing your page to get the latest changes.
